expression/builtin_string.go

```diff
@@ -540,33 +534,39 @@ type upperFunctionClass struct {
 }
 
 func (c *upperFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
-	sig := &builtinUpperSig{newBaseBuiltinFunc(args, ctx)}
+	retType := oneArgsInferType(args)
+	bf, err := newBaseBuiltinFuncWithTp(args, retType, ctx, tpString)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	sig := &builtinUpperSig{baseStringBuiltinFunc{bf}}
 	return sig.setSelf(sig), errors.Trace(c.verifyArgs(args))
 }
 
+func oneArgsInferType(args []Expression) *types.FieldType {
```

Comment:
This is not a good function name. You can name it `InferTypeStringString` instead, which be used for all functions that only have a string argument and a string return value.